shell: Add shell_exec and shell tests

Vudentz opened the pull request:
This adds shell_exec which can be used to execute commands directly
without the use of a console which is useful for both testing as well
as interfacing with applications/upper layer which would like to have
access to shell commands directly.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
shell: Add shell_exec

This adds shell_exec which can be used to execute commands directly
without the use of a console which is useful for both testing as well
as interfacing with applications/upper layer which would like to have
access to shell commands directly.

In addition to that this may be more trivial to interface with instead
of using fifos like uart_register_input and telnet_register_input do.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jukkar reviewed (commented):
@pfalcon Would it make sense in micropython to use this instead of calling the net-shell C-functions directly?

jukkar reviewed (dismissed):
I like this approach so +1 from me.

jhedberg commented:
A couple of comments:

Make it super clear in the documentation that the shell commands are not a formal API or something that can be considered stable. I.e. anyone writing code that uses the command strings should be prepared for that code to stop working at some point.

Secondly, make it clear in the documentation that C string literals cannot be passed to this function because it modifies the input (by inserting nul characters to terminate each command line argument). Instead, the caller will need to do something like "char cmd[] = "my command"; shell_exec(cmd);"

nashif reviewed (changes requested):
Please provide a testcase for this new usage and document this in the shell subsystem documentation.

Vudentz committed:
shell: Add define to _SHELL_H_

This prevents the header to be included multiple times.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
shell: Return proper errors if the command fails

If the command cannot be execute code should return a proper since this
may not be a user input.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
shell: Adjust arguments when executing a command of different module

The command callback might not recognize commands if the input comes
with the module name as first parameter as both argc and argv will be
off by one.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
net: shell: Remove code adjust arguments

The shell takes care of removing the module name so it is no longer
necessary to have this adjustment.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
tests: Add shell tests

This adds the following tests:

help tests:
shell_exec(help): 0
shell_exec(help dummy): 0
shell_exec(help invalid): -22

select tests:
shell_exec(select): 0
shell_exec(select dummy): 0
shell_exec(select invalid): -22

module tests:
shell_exec(dummy cmd1): 0
shell_exec(dummy cmd1 arg1): -22
shell_exec(dummy cmd2 arg1): 0
shell_exec(dummy cmd2 arg1 arg2): -22
shell_exec(dummy cmd3 arg1 arg2): 0
shell_exec(dummy cmd3 arg1 arg2 arg3): -22
shell_exec(dummy cmd4 arg1 arg2 arg3): -22
shell_exec(cmd1): 0
shell_exec(cmd1 arg1): -22
shell_exec(cmd2 arg1): 0
shell_exec(cmd2 arg1 arg2): -22
shell_exec(cmd3 arg1 arg2): 0
shell_exec(cmd3 arg1 arg2 arg3): -22
shell_exec(cmd4 arg1 arg2 arg3): -22

app handler tests:
shell_exec(cmd4 arg1 arg2 arg3): 0

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jukkar reviewed (dismissed):
LGTM

Vudentz commented:
@nashif Tests are now included.

nashif commented on tests/shell/testcase.ini:
```diff
@@ -0,0 +1,4 @@
+[test]
+build_only = true
```
remove, this should be run in qemu during sanitycheck

Vudentz replied:
Done

nashif commented on tests/shell/src/main.c:
```diff
@@ -0,0 +1,132 @@
+ *  @brief Interactive shell test suite
+ *
+ */
+
```
put this below the license header

Vudentz replied:
Done

nashif commented on tests/shell/src/main.c:
```diff
@@ -0,0 +1,132 @@
+
+static void test_help(void)
+{
+	TC_PRINT("\nhelp tests:\n");
```
this duplicates what the testsuite already prints, too verbose. I would remove all those prints, thats how it looks now:

`tc_start() - test_select

select tests:`

Vudentz replied:
Done

nashif commented on tests/shell/src/main.c:
```diff
@@ -0,0 +1,132 @@
+
+static void test_select(void)
+{
+	TC_PRINT("\nselect tests:\n");
```
same, remove this, this is already printed by ztest, other tests below as well.

Vudentz replied:
Done

nashif reviewed (changes requested):
Please move to test to:

tests/subsys/shell

pfalcon reviewed (approved):
>@pfalcon Would it make sense in micropython to use this instead of calling the net-shell C-functions directly?

As discussed in the previous patch by @Vudentz, using IP stack diagnostic functions is a more direct way. We don't have such functions well defined, but un-static-izing functions used by net shell was a first step in that direction.

These changes are definitely good improvements to the shell system, and yep, if direct access to net subsystem functions will be closed, can serve as a (somewhat hackish) fallback for that too.

Vudentz commented:
> As discussed in the previous patch by @Vudentz, using IP stack diagnostic functions is a more direct way. We don't have such functions well defined, but un-static-izing functions used by net shell was a first step in that direction.

> These changes are definitely good improvements to the shell system, and yep, if direct access to net subsystem functions will be closed, can serve as a (somewhat hackish) fallback for that too.

Im not directly targeting any changes to net_shell.h right now, though this offer an alternative to having the callback handler exposed in headers like that I guess we need a proper discussion on what are the benefit vs shell_exec.
